### ISSUE TITLE
How to add applications so they show up in Applications (or Launchpad search)

### ISSUE THREAD

**Issue by spease** [LnL7/nix-darwin#214]
It seems if I add applications in the user configuration file, they end up in ~/Applications rather than in the system Applications. I there a way to add packages to the system so the symlinks will show up in the root Applications, or at least will be searchable with Launchpad?

I couldn't find any "Getting Started" documentation with nix-darwin that would answer this question, but I'm very interested in getting more hands-on experience with the nix ecosystem

**Comment by jsravn** [comment]
@spease It's currently bugged in nix-darwin. There is a workaround at https://github.com/LnL7/nix-darwin/issues/139#issuecomment-666771621. Hopefully it will get a fix soon - it should be easy to make a PR for it.

**Comment by purcell** [comment]
Note that the linked comment is about making the apps show up in LaunchPad etc., but they still live in `~/Applications`. There was other discussion in that issue about the wider question of `/Applications` vs. `~/Applications`, and a corresponding PR was opened (#153) and later abandoned.

**Comment by spease** [comment]
I’m not as worried about that particular aspect, it’s more having a GUI app behave ergonomically once it’s included in the environment. I’m installing things for my user profile so far, not the system, so it makes sense to me that they show up in ~ instead of /.

**Comment by LnL7** [comment]
Yeah, moving the system applications out of the home folder would be good.

**Comment by wangkev** [comment]
I applied the mentioned patch https://github.com/LnL7/nix-darwin/issues/139#issuecomment-666771621, and can see `~/Applications` with the `.apps`. Spotlight and Launchpad still do not see the apps though. Is that expected? I'm on Catalina if that makes any difference.

**Comment by tonylotts** [comment]
> I applied the mentioned patch [#139 (comment)](https://github.com/LnL7/nix-darwin/issues/139#issuecomment-666771621), and can see `~/Applications` with the `.apps`. Spotlight and Launchpad still do not see the apps though. Is that expected? I'm on Catalina if that makes any difference.

Some applications in `~/Applications` (my own derivations) are listed in LaunchPad; however to get Emacs.app to show up in LaunchPad, I had to create the alias in `/Applications`.  I think there must be some metadata that is different WRT Emacs.app, as opposed to other application bundles.

**Comment by webframp** [comment]
I would definitely be in favor of some method for linking system installed GUI apps into `/Applications` instead of `~/Applications`. Finder, Alfred etc can't find symlinks in `~/Applications` even if they can be made to show up in Launchpad

**Comment by leira** [comment]
I have the same issue. There is a symlink of `Nix Apps` under `~/Applications`. In `Nix Apps` linked folder, there is a `Kitty.app` symlink. Spotlight doesn't show Kitty.

I did some experiment, it seems the problem is the symlinks. If I copy the `Kitty.app` file into `~/Applications`, it works. Even if I put the `Kitty.app` symlink in `/Applications`, Spotlight still doesn't see the app. It seems Spotlight simply doesn't support symlinks.

I'm not sure what is the difference between symlinks and aliases in macOS. But even I tried with aliases, Spotlight doesn't see it either.

**Comment by IvarWithoutBones** [comment]
Folders symlinked to `~/Applications` do not show up in spotlight, and copying them seems like a waste of disk space. Not to mention that it can take a while too if you have a lot of packages. I found a workaround using APFS aliases, those show up in spotlight and barely take any disk space.

 This is my activation script:

```nix
  # Nix-darwin does not link installed applications to the user environment. This means apps will not show up
  # in spotlight, and when launched through the dock they come with a terminal window. This is a workaround.
  # Upstream issue: https://github.com/LnL7/nix-darwin/issues/214
  system.activationScripts.applications.text = lib.mkForce ''
    echo "setting up ~/Applications..." >&2
    applications="$HOME/Applications"
    nix_apps="$applications/Nix Apps"

    # Needs to be writable by the user so that home-manager can symlink into it
    if ! test -d "$applications"; then
        mkdir -p "$applications"
        chown ${username}: "$applications"
        chmod u+w "$applications"
    fi

    # Delete the directory to remove old links
    rm -rf "$nix_apps"
    mkdir -p "$nix_apps"
    find ${config.system.build.applications}/Applications -maxdepth 1 -type l -exec readlink '{}' + |
        while read src; do
            # Spotlight does not recognize symlinks, it will ignore directory we link to the applications folder.
            # It does understand MacOS aliases though, a unique filesystem feature. Sadly they cannot be created
            # from bash (as far as I know), so we use the oh-so-great Apple Script instead.
            /usr/bin/osascript -e "
                set fileToAlias to POSIX file \"$src\" 
                set applicationsFolder to POSIX file \"$nix_apps\"
                tell application \"Finder\"
                    make alias file to fileToAlias at applicationsFolder
                    # This renames the alias; 'mpv.app alias' -> 'mpv.app'
                    set name of result to \"$(rev <<< "$src" | cut -d'/' -f1 | rev)\"
                end tell
            " 1>/dev/null
        done
  '';
```
I have done something similar for home-manager, which integrates with this just fine.

Commit from my dotfiles: https://github.com/IvarWithoutBones/dotfiles/commit/0b3faad8bd1d0e1af6103caf59b206666ab742f4

**Comment by hacker1024** [comment]
> Folders symlinked to ~/Applications do not show up in spotlight, and copying them seems like a waste of disk space. Not to mention that it can take a while too if you have a lot of packages.

What about taking advantage of APFS's copy-on-write functionality to make a clone of the app bundle? That should avoid wasting resources, and it's as easy as `cp -c`.

The cloned bundles will have to be deleted and recreated before each update. This may cause issues with apps that store user data within their bundles, though, which is bad practice and uncommon.

**Comment by cdmistman** [comment]
@IvarWithoutBones's solution worked for applications installed with `environment.systemPackages`, but unfortunately the home-manager version of the script only works on previous home-manager generations' `~/Applications/Home Manager Apps`.

I just spent most of my Sunday experimenting with the various already-posted solutions, if anybody wants to catch up on it you can see [my comment here](https://github.com/nix-community/home-manager/issues/1341#issuecomment-1445539742)

**Comment by jacekszymanski** [comment]
Hi, 

I have encountered another issue: when the apps are linked, as now do both nix-darwin and home-manager, file associations don't get updated on switch. That is, when I right-click-open a file with an application, which is linked from the generation X, then switch, and then again right-click-open the same file with the same-name application, the instance from the generation X is picked up, not from the generation X+1. It changes only after manually opening the X+1-generation app.

**Comment by vs49688** [comment]
> ```nix
>  # Spotlight does not recognize symlinks, it will ignore directory we link to the applications folder.
>             # It does understand MacOS aliases though, a unique filesystem feature. Sadly they cannot be created
>             # from bash (as far as I know), so we use the oh-so-great Apple Script instead.
> ```

Bit late to the party, can confirm using APFS aliases works. You can avoid AppleScript with a tiny ObjC helper:

https://github.com/vs49688/mkalias/blob/master/mkalias.m

**Comment by pshirshov** [comment]
So, these two snippets work

```
  # see https://github.com/LnL7/nix-darwin/blob/master/modules/system/activation-scripts.nix
  system.activationScripts.postUserActivation.text = ''
    app_folder="$HOME/Applications/Nix Trampolines"
    rm -rf "$app_folder"
    mkdir -p "$app_folder"
    for app in $(find "${config.system.build.applications}/Applications" -type l); do
        app_target="$app_folder/$(basename $app)"
        echo "mkalias \"$app\" \"$app_target\"" >&2
        ${pkgs.mkalias}/bin/mkalias "$app" "$app_target"
    done
  '';
```

home-manager:

```
  # don't forget to add 'com.apple.alias-file' without quotes into alfred's Extras
  home.activation = {
    aliasHomeManagerApplications = lib.hm.dag.entryAfter [ "writeBoundary" ] ''
      app_folder="${config.home.homeDirectory}/Applications/Home Manager Trampolines"
      rm -rf "$app_folder"
      mkdir -p "$app_folder"
      for app in $(find "$genProfilePath/home-path/Applications" -type l); do
          app_target="$app_folder/$(basename $app)"
          echo "mkalias \"$app\" \"$app_target\"" >&2
          $DRY_RUN_CMD ${pkgs.mkalias}/bin/mkalias "$app" "$app_target"
      done
    '';
  };
```

Although there are at least two issues:

1. Permissions (e.g. Full Disk Access) do not survive package version updates because they are given directly to the apps in nix store
2. Spotlight depresses the search results and Alfred  needs additional configuration to even make the results visible.

**Comment by pshirshov** [comment]
There is a supposedly better version of my snippet: https://github.com/nix-community/home-manager/issues/1341#issuecomment-2049723843

**Comment by pshirshov** [comment]
Another approach (https://gist.github.com/berryp/7af4a1f534027807b26fff706a61b1ce):

```nix
  home.activation = {
    aliasHomeManagerApplications = lib.hm.dag.entryAfter [ "writeBoundary" ] ''
      rsyncArgs="--archive --checksum --chmod=-w --copy-unsafe-links --delete"
      apps_source="$genProfilePath/home-path/Applications"
      app_target="${config.home.homeDirectory}/Applications/Home Manager Trampolines"
      rm -rf "$app_target"
      mkdir -p "$app_target"
      ${pkgs.rsync}/bin/rsync $rsyncArgs "$apps_source" "$app_target"
    '';
  };
```

```nix
  system.activationScripts.postUserActivation.text = ''
    rsyncArgs="--archive --checksum --chmod=-w --copy-unsafe-links --delete"
    apps_source="${config.system.build.applications}/Applications"
    app_target="$HOME/Applications/Nix Trampolines"
    rm -rf "$app_target"
    mkdir -p "$app_target"
    ${pkgs.rsync}/bin/rsync $rsyncArgs "$apps_source" "$app_target"
  '';
```

**Comment by remi-gelinas** [comment]
Is there a reason it's suggested to setup trampolines to systems application bundles in the user's home directory, rather than copying/linking them to `/Applications`?

**Comment by cdmistman** [comment]
> Is there a reason it's suggested to setup trampolines to systems application bundles in the user's home directory, rather than copying/linking them to `/Applications`?

trampolines are better than copying because they don't use the full storage of the application (as mentioned, this can be in the GBs), while they're better than linking because Spotlight doesn't index links in `/Applications`

**Comment by jpsecher** [comment]
For Home Manager, I had to do this:
```
 system.activationScripts.postUserActivation.text = ''
   apps_source="$HOME/Applications/Home Manager Apps"
   moniker="Nix Trampolines"
   app_target_base="$HOME/Applications"
   app_target="$app_target_base/$moniker"
   mkdir -p "$app_target"
   ${pkgs.rsync}/bin/rsync --archive --checksum --chmod=-w --copy-unsafe-links --delete "$apps_source/" "$app_target"
 '';
```

**Comment by pshirshov** [comment]
Almost all the approaches, including trampolines, are affected by this: https://github.com/hraban/mac-app-util/issues/19

**Comment by gshpychka** [comment]
> For anyone, like me, who's reading this looking for solutions:
> 
> You _can_ use nix-darwin's [homebrew module](https://github.com/nix-darwin/nix-darwin/blob/master/modules/homebrew.nix). It installs only the listed casks and uninstalls anything you delete from the config.
> 
> GUI apps installed with homebrew will be installed to /Applications

100%, I went this route never looked back, stopped thinking about the problem a long time ago. All GUI apps are managed via nix-darwin declaratively, and even homebrew itself is managed declaratively via nix-homebrew

**Comment by dwt** [comment]
@mikhail5555 I recommend checking out and supporting the work @Samasaur1 is doing with his files module, that will eventually enable us to have the same level of control for GUI applications installed via nix.

**Comment by Samasaur1** [comment]
@dwt's PR (#1396) will solve this issue by copying apps into `/Applications`. There's no perfect solution, but this seems like the one with the least drawbacks. All the workarounds in this issue should continue to work, though, if your circumstances need you to prioritize saving space.

**Comment by Enzime** [comment]
As of #470, Applications are now linked into `/Applications` instead of `~/Applications`

**Comment by poopsicles** [comment]
> What I don't like is how updates are handled - those are not declarative, unfortunately.
@gshpychka if you set `mutable.taps` to `false` and set `homebrew.taps` to `builtins.attrNames config.nix-homebrew.taps`, as well as `homebrew.cleanup` to `zap`, would `brew update` do anything?

i just tried it and it gives me a Permission denied error--the taps are now only managed by the flake inputs--i think that's pretty declarative?

**Comment by gshpychka** [comment]
> > What I don't like is how updates are handled - those are not declarative, unfortunately.
> > [@gshpychka](https://github.com/gshpychka) if you set `mutable.taps` to `false` and set `homebrew.taps` to `builtins.attrNames config.nix-homebrew.taps`, as well as `homebrew.cleanup` to `zap`, would `brew update` do anything?
> 
> i just tried it and it gives me a Permission denied error--the taps are now only managed by the flake inputs--i think that's pretty declarative?

Yeah good point. Except for the apps that update themselves.

**Comment by jennifgcrl** [comment]
macos 26 seems to understand symlinks in /Applications and ~/Applications

Edit: Sorry, I'm not actually sure this is true. I think symlinked apps didn't show up in Alfred prior to 26 but now do, whereas Spotlight/Launchpad still doesn't show them.

**Comment by Samasaur1** [comment]
and i got so excited, too! :)

Yeah, a couple of us in the nix on macOS matrix room checked this behavior as well and it does not seem to have changed. A shame, because they're doing such a big Spotlight rework already.

**Comment by nevir** [comment]
One alternative avenue to consider is to make the `/nix` volume Spotlight-indexable (e.g. via `mdutil`); and to make sure that the correct nix paths are indexed (e.g. via `mdimport`)

The first thing I run into when attempting that approach is that I am unable to turn spotlight on for the `/nix` volume (as configured by the Determinate installer):

```
> sudo mdutil -i on /nix
/nix:
Error: unable to perform operation.  (-400)
        Error: unknown indexing state.
```

Also, weirdly:

```
> sudo rm -rf /nix/.Spotlight-V100
> sudo mkdir /nix/.Spotlight-V100 
mkdir: /nix/.Spotlight-V100: Operation not permitted
```

IIRC this might be due to the nix volume being encrypted and spotlight services needing to access the volumes directly (not as the active user)?

---

Update: See #1576 for more investigation

**Comment by jennifgcrl** [comment]
@Samasaur1 I filed a bug ticket with Apple and it looks like they fixed it.

<img width="596" height="613" alt="Image" src="https://github.com/user-attachments/assets/ed0d120e-65db-485d-9f97-6b382e76bb49" />

**Comment by nhooey** [comment]
Amazing! Thanks so much!
How did you get Apple to actually reply to, and then also fix a bug?

> [@Samasaur1](https://github.com/Samasaur1) I filed a bug ticket with Apple and it looks like they fixed it.

**Comment by dwt** [comment]
It happens, just very rarely.

**Comment by Samasaur1** [comment]
okay I'm testing in a Tahoe VM...

Assuming Apple actually did fix this in a way that works for us, we'll have to figure out what nix-darwin's plan is moving forward, but I suspect we'll stay with copying for now

**Comment by Samasaur1** [comment]
forgot to update here but they did not fix the problem, at least when the source file is in the nix store. we'll be sticking with copying for the time being
